### ISSUE TITLE
The godot class heirarchy doesn't match how regular javscript classes work.

### ISSUE THREAD

**Issue by TerranceN** [godotjs/GodotJS#12]
...which is understandable, I'm amazed all of of this works so well out of the box, fantastic work!

But anyway, the typescript definitions say that Node3D extends Node, and yet:

```typescript
Node3D.prototype instanceof Node
(Node3D.prototype as any).__proto__ === Node.prototype
```

are both `false`, despite these normally being true in javascript:

```typescript
class A {};
class B extends A {};
B.prototype instanceof A // is true
(B.prototype as any).__proto__ === A.prototype // is true
```

This means something like making an export function with default `export_` arguments using `reflect-metadata` like:
```typescript
export function Export<
  Key extends string,
  Type extends Node & {[key in Key]: Resource | Node}
>(): (target: Type, key: Key) => void;
export function Export(...args: Parameters<typeof export_>): (target: Node, key: string) => void;
export function Export(...args: any[]) {
  return (
    target: Node,
    memberName: string
  ) => {
    const type = Reflect.getMetadata("design:type", target, memberName);
    // If args were specified, use those
    if (args[0]) {
      return export_(args[0], args[1])(target, memberName);
    } else {
      // Otherwise automatically determine it
      if (type === Resource || type.prototype instanceof Resource) {
        return export_(
          Variant.Type.TYPE_OBJECT,
          {
            hint: PropertyHint.PROPERTY_HINT_RESOURCE_TYPE,
            hint_string: type.name
          }
        )(target, memberName)
      } else if (type === Node || type.prototype instanceof Node) {
        return export_(
          Variant.Type.TYPE_OBJECT,
          {
            hint: PropertyHint.PROPERTY_HINT_NODE_TYPE,
            hint_string: type.name,
          }
        )(target, memberName)
      }
    }

    // Typescript should ensure we never hit this, but error just in case
    throw new Error(`Failed to Export type ${type.name}`)
  }
}
```

won't actually work because the `type.prototype instanceof Resource` and `type.prototype instanceof Node` will return false for any subclasses of those types, unlike normal javascript class hierarchies.

Also this fake class heirarchy means we can't "polyfill" godot base classes with additional convenience methods for subclasses. A polyfill for Node3D will work on scripts that extend Node3D, but not on scripts that extend a type that extends Node3D, like CharacterBody3D.

i.e.

```typescript
(Node3D.prototype as any).test = "does this work?"
(CharacterBody3D.prototype as any).test // is undefined
```

doesn't work, but it does in regular javascript:

```typescript
class A {};
class B extends A {};
(A.prototype as any).test = "does this work?"
(B.prototype as any).test // is "does this work?"
```

In addition to generating the types with an inheritance hierarchy, could we also generate the actual classes (wherever that happens) with inheritance to match? Looking through the code here, I would have thought [this](https://github.com/godotjs/GodotJS/blob/main/bridge/jsb_object_bindings.cpp#L128-L136) does exactly that, but maybe I'm looking in the wrong place?

I figured out how to compile with VeryVerbose and run with `--verbose` and the JSB_LOG there is firing. Also thought this might be a v8 binding issue, so I tried compiling with `use_quickjs=yes`, but I have the same issue.

I see `expose_godot_object_class` seems to grab classes from `native_classes_` if they're a godot class (at least I think?). Maybe those are different than the v8 classes somehow? I.e. `Node3D` extends a different `Node` than the one we can access from javascript? At least that's my hunch, because printing `Object.keys((Node.prototype as any).__proto__)` prints all the methods of a godot `Object`, so it's inheriting from _something_, just seemingly not the `Object` I can import from `"godot"`.

This is all very new to me, sorry if I'm missing something obvious!

**Comment by TerranceN** [comment]
Whoops, I didn't realize when compiling with quickjs I also set `dev_build=true`, which changed the output binary name, so I was still running the v8 build when I tested this.

Switching to quickjs fixes these issues for me, so this seems to be a v8-specific issue.

**Comment by ialex32x** [comment]
Thanks for reporting it with such detailed info. The prototype issue has driven me crazy for a long time 😂, I'll check it out right away.

**Comment by ialex32x** [comment]
After dozens of attempts, it was finally fixed (56bb8ac). The problem was caused by incorrectly registering an ObjectTemplate into a FunctionTemplate (a nested enum within a class).

**Comment by TerranceN** [comment]
Amazing work, thank you!
